FEAT: Menu element plugin for page builder [SECOND PR]

TommyJackson85 opened the pull request:
## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
Closes #1176

## Your solution

THIS IS A NEW PR IN REPLACEMENT TO [PREVIOUS PR #1207](https://github.com/webiny/webiny-js/pull/1207) DUE TO A SEVERE BUG.

Creating a menu plugin package used for the app page builder package.

Creating element plugin within the app page builder.

<!--- Please describe your solution, have you encountered any issues along the way? -->

## How Has This Been Tested?
In progess.
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if relevant):
None yet.

TommyJackson85 committed:
feat: added copy of editor menu folder

TommyJackson85 committed:
feat: exporting plugins from menus respect
ive parent render and editor folders

TommyJackson85 committed:
feat: copied renders menu element folders and files

TommyJackson85 committed:
feat: unpdated app page builder types to version from last pr

TommyJackson85 commented:
Previous PR had a bug which set me back on fixing this issue. This is a fresh PR with the most updated files of Webiny plus the relevant changes for a Menu element from the [previous PR's](https://github.com/webiny/webiny-js/pull/1207) commit e1e3776 (not all changes).
Previous PR is to be used for reference only.

TommyJackson85 committed:
feat: removed comments from graphql to solve bug

TommyJackson85 committed:
feat: updated package json to most recent version

TommyJackson85 committed:
feat: fixed package json error

TommyJackson85 committed:
feat: removed filter options and amount per page options

TommyJackson85 committed:
feat: menu element now replicated form adder for editor files

TommyJackson85 committed:
feat: value in options form updates correctly

TommyJackson85 committed:
feat: menu listings display in form but not rendered in editor on select

TommyJackson85 committed:
feat: rendering menu form template to editor on drop to display data

TommyJackson85 committed:
feat: removed comments console logs, renamed console logs

TommyJackson85 committed:
feat: rendering menu form template on preview screens

TommyJackson85 committed:
feat: removed repeated use query, console logs and comments

TommyJackson85 committed:
feat: unknown change

TommyJackson85 committed:
feat: renamed variablesrewrote logs in editor folder for further clarity

TommyJackson85 commented:
@Pavel910 What do you think of my work so far?
![tutorial-1](https://user-images.githubusercontent.com/22279028/95266510-34f4b380-082b-11eb-836c-60771c9bfcd2.gif)

Take a look at the these files too. These are the files that I have been working on. Let me know if anything I wrote is a little confusing.

***Editor files:***
packages/app-page-builder/src/editor/plugins/elements/menu/index.tsx
packages/app-page-builder/src/editor/plugins/elements/menu/Menu.tsx
packages/app-page-builder/src/editor/plugins/elements/menu/MenuDesignSettings.tsx
packages/app-page-builder/src/editor/plugins/elements/menu/graphql.ts

***Render files:***
packages/app-page-builder/src/render/plugins/elements/menu/index.tsx
packages/app-page-builder/src/render/plugins/elements/menu/Menu.tsx
packages/app-page-builder/src/render/plugins/elements/menu/components/MenuTemplate.tsx
packages/app-page-builder/src/render/plugins/elements/menu/graphql.ts

***types:***
packages/app-page-builder/src/types.ts

TommyJackson85 committed:
feat: removed grid menuimport which caused a bug in build

brunozoric commented on packages/app-page-builder/src/editor/plugins/elements/menu/Menu.tsx:
```diff
@@ -0,0 +1,84 @@
+
+const Menu = (props: MenuElementProps) => {
+    //menu data passed through here
+    const { element, isActive } = props;
```
You can use {element, isActive} directly as function argument

brunozoric commented on packages/app-page-builder/src/editor/plugins/elements/menu/Menu.tsx:
```diff
@@ -0,0 +1,84 @@
+    console.log("MENU PROPS ::::;;;;");
+    console.log(props);
+ 
+    const data = get(element, "data") || {};
```
Use optional instead of get if you are unsure that element actually is a object. So it looks like this: `const data = element?.data || {}`;
From MenuElementProps it is notable that element must be a PbElement and it has data property. So, basically, it would be enough to write `const data = element.data;`.

TommyJackson85 replied:
@brunozoric Im sure its an object, so I am just going to call get and remove ```|| {} ;```

brunozoric commented on packages/app-page-builder/src/editor/plugins/elements/menu/Menu.tsx:
```diff
@@ -0,0 +1,84 @@
+    } else {
+        return <div>Selected menu component not found!</div>;
+    }
+    console.log(`menu ID: ${menuId}`);
```
For this part, i would write it totally different since you do not care about menuId yet, eg:
```
if (!data?.settings) {
  return <div>Selected menu component not found!</div>
}
```

brunozoric commented on packages/app-page-builder/src/editor/plugins/elements/menu/Menu.tsx:
```diff
@@ -0,0 +1,84 @@
+        };
+        render = <MenuComponent {...props} />;
+    }
+
```
Here, a render var would actually be a function, eg:
```
const render = (menuId?: string) => {
  if (!menuId) {
    return <span>Menu not selected.</span>;
  }
  return <MenuComponent menu={menuId} preview={true} />;
};
```
And then just pass the value that should be menuId:
```
render(data?.settings?.menu?.element);
```

brunozoric commented on packages/app-page-builder/src/render/plugins/elements/menu/components/MenuItems/findObject.ts:
```diff
@@ -0,0 +1,26 @@
+/**
+ * Recursively search for an object with given ID in the given source array.
+ */
+const findObject = (source: Array<any>, id: string): any => {
```
This is ok, but let's try to avoid lodash when possible :)
Here is how it would look with raw JS:
```
type FindObjectResult<T> = {
  source: T[],
  index: number,
  item: T,
};
const findObject = <T extends any>(source: T[], id: string): FindObjectResult<any> | null => {
  if (source.length === 0) {
    return null;
  }
  const index = source.findIndex(item => item.id === id);
  if (index >= 0) {
    return {source, index, item: source[index]};
  }
  return source.reduce((acc: FindObjectResult<T> | null, item: T) => {
    if (!item.children) {
      return acc;
    }
    return findObject(item.children, id);
  }, null);
};
```

Didn't run the fn but it should work :)

brunozoric reviewed (commented):
Looks good overall (didn't run it) and I left you some comments on what to improve and refactor.

TommyJackson85 committed:
feat: menu id now passed through a new render function for editor

TommyJackson85 committed:
feat: simpified get options function in menu designed settings

TommyJackson85 committed:
feat: removed empty string and move preview box styler outside export

TommyJackson85 committed:
feat: render in render menu  now a function and passes in the menu id

TommyJackson85 committed:
feat: render in render menu now a function and passes in the menu id

TommyJackson85 committed:
feat: removed colons

TommyJackson85 committed:
feat: menu template now displays title

TommyJackson85 committed:
feat: renamed menu design settings file to menu options settings

TommyJackson85 commented:
@brunozoric I basically copied the menu form from from the menu builder page on the webiny project (http://localhost:3001/page-builder/menus) but I am not sure if this is what we want to have displayed. 
Is the menu meant to be used for nav bars and side navs with it's menu items? Is there other uses for them?
This is whats displayed when rendering the page.

<img width="1436" alt="Screenshot 2020-10-11 at 21 47 38" src="https://user-images.githubusercontent.com/22279028/95690001-a9a16680-0c0c-11eb-8aff-a87769019c3a.png">
<img width="922" alt="Screenshot 2020-10-11 at 21 47 57" src="https://user-images.githubusercontent.com/22279028/95690004-ad34ed80-0c0c-11eb-99f8-1c56d3f2ffb0.png">

TommyJackson85 committed:
feat: added component form option and menu navigation component file

TommyJackson85 commented:
@brunozoric I added a new option input for the user to choose a component to render the menu data. I also added a new render plugin which will be used for a Menu Navigation component with dropdown functions but its still being worked on. I intend to remove the Menu Form component option later as its not relevent. I wanted to show a gif display of how it the form currently looks but theres a separate bug from the master branch causing it to break, in which just posted about on the slack channel.

TommyJackson85 commented:
@brunozoric Form now has a component input field to choose a component to render menu with. Still working on Menu Navigation which currently only displays a string as you can see below. I intend to remove Menu Form as I feel it is not relevant for this task but I am keeping it for testing for the time being.
![tutorial-1](https://user-images.githubusercontent.com/22279028/96508564-b3514c80-1252-11eb-9252-ac675fa77f8b.gif)

TommyJackson85 committed:
feat: react now imported to menu navigation file

TommyJackson85 committed:
feat: renamed files and menu rendering catches missing menu ids

TommyJackson85 committed:
feat: menu navigation currently a plain list with no styling

TommyJackson85 committed:
feat: removed menu form component option

TommyJackson85 committed:
feat: removed comments logs unused files in prep for v5 migrations

TommyJackson85 committed:
feat: changed options labels and passing theme to menu component

Pavel910 commented:
v5 changed the whole editor API, so I'm closing this and we'll come back to this topic later.
